main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -492,13 +493,49 @@ object Inliner {
     * @param exps the arguments to the function.
     * @param ctx0 the local context.
     */
-  private def shouldInlineDef(defn: MonoAst.Def, exps: List[Expr], ctx0: LocalContext)(implicit sctx: SharedContext): Boolean = {
-    !sctx.delta.contains(defn.sym) &&
-      !ctx0.currentlyInlining &&
-      !defn.spec.defContext.isSelfRef &&
+  private def shouldInlineDef(defn: MonoAst.Def, exps: List[Expr], ctx0: LocalContext)(implicit sym0: Symbol.DefnSym, sctx: SharedContext): Boolean = {
+    if (defn.spec.ann.isDontInline) {
+      return false
+    }
+
+    if (sctx.delta.contains(defn.sym)) {
+      return false
+    }
+
+    if (defn.spec.ann.isInline) {
+      if (defn.sym == sym0) {
+        return false
+      }
+
+      if (defn.spec.defContext.isSelfRef) {
+        return atomicallyCheckShouldInlineRecursive(defn)
+      }
+
+      return true
+    }
+
+    !defn.spec.defContext.isSelfRef && !ctx0.currentlyInlining &&
       (isSingleAction(defn.exp) || isSimple(defn.exp) || hasKnownLambda(exps))
   }
 
+  private def atomicallyCheckShouldInlineRecursive(defn: MonoAst.Def)(implicit sym0: Symbol.DefnSym, sctx: SharedContext): Boolean = {
```

Comment:
What is all this? I just want it to inline.

Reply:
`@Inline` should simply mean: always inline when possible - i.e. disable any checks.

Reply:
We still only want to unroll a recursive function once at a call site.

Reply:
Not something that requires more complexity.